src/scheduler/Scheduler.cpp

```diff
@@ -296,8 +311,20 @@ faabric::util::SchedulingDecision Scheduler::makeSchedulingDecision(
                 int available = r.slots() - r.usedslots();
                 int nOnThisHost = std::min(available, remainder);
 
-                for (int i = 0; i < nOnThisHost; i++) {
-                    hosts.push_back(h);
+                // Under the pairs topology hint, we never allocate a single
```

Comment:
My gut here would instead be:

```cpp
if(topologyHint == faabric::util::SchedulingTopologyHint::PAIRS &&
      nOnThisHost < 2) {
    // Move on if we can't colocate function with at least one other
    continue;
}
```

Reply:
I don't think this behaves as we expect it. For example:
* We have 4 hosts with 4 slots each.
* We want to schedule 9 requests (with the `NEVER_ALONE` hint).
* We expect 4 requests scheduled to the first host, and 5 scheduled to the latter.
However, I think your solution will schedule 5 requests on the first host and 4 on the second, as it would exhaust all possible hosts (`nOnThisHost == 1` for all of them), and resort to overload the master.